### PR TITLE
feat(meshservice): pure label compute helpers

### LINKED ISSUE #16545: Phase 3 — Pure label-compute helpers (dpContribution, mergeAcrossDataplanes)
Parent: #16542

Add two pure label-compute helpers. **Fully testable in isolation**; no wiring yet (Phase 4 wires them in).

## T3.1 — `dpContribution` helper (per-DP, intra-DP consensus + DP labels)

**File:** `pkg/core/resources/apis/meshservice/generate/labels.go` (new) or appended to `generator.go`.

```go
// dpContribution returns this Dataplane's contribution to a MeshService's labels:
// cross-inbound consensus on non-reserved tags (drop on intra-DP disagreement),
// overlaid by non-reserved DP resource labels.
// inbounds: the inbounds in this DP backing the MeshService (sharing kuma.io/service).
func dpContribution(
    dp *core_mesh.DataplaneResource,
    inbounds []*mesh_proto.Dataplane_Networking_Inbound,
    allowSet map[string]struct{}, // nil = all non-reserved allowed
    metric prometheus.Counter,    // dropped-labels counter
    log logr.Logger,
) map[string]string {
    // 1. Cross-inbound consensus on tags. Drop keys whose value disagrees across inbounds.
    // 2. Overlay DP resource labels — these beat inbound tags within a single DP.
    // Skip+log+metric any key/value failing K8s qualified-name / label-value validation.
}
```

Use `validation.IsQualifiedName` and `validation.IsValidLabelValue` from `k8s.io/apimachinery/pkg/util/validation` (already imported elsewhere in the project).

Use `mesh_proto.IsReservedLabelKey` (from Phase 1).

## T3.2 — `mergeAcrossDataplanes` helper (per-key majority-wins, newest tiebreak)

```go
// mergeAcrossDataplanes returns the merged label set for a MeshService that's
// backed by the given Dataplanes. Per-key majority-wins; ties broken by newest
// DP creation time; further ties broken by lexicographic value sort.
// contribution(dp) returns this DP's Stage-1 label contribution.
func mergeAcrossDataplanes(
    dps []*core_mesh.DataplaneResource,
    contribution func(*core_mesh.DataplaneResource) map[string]string,
) map[string]string {
    // For each key K across all DPs:
    //   tally distinct values, recording max creation time per value
    //   pick max-count value; tie → newest creation time; tie → lex sort
    // A key appears in output iff at least one DP defines it.
}
```

Note: `core_mesh.SortDataplanes` already sorts by creation time then name (`pkg/core/resources/apis/mesh/dataplane_helpers.go:460-468`) — reuse if you need a deterministic iteration order.

## Tests

**File:** `pkg/core/resources/apis/meshservice/generate/labels_test.go` (new).

Use `DescribeTable` + `Entry` style. Cases:

**Per-DP (`dpContribution`):**
1. Single inbound, custom tag `appci=jeffy` → propagated.
2. DP resource label `color=blu` → propagated.
3. DP label `color=blu` AND inbound tag `color=red` → output `color=blu` (DP wins).
4. Two inbounds backing the service, both `appci=jeffy` → propagated.
5. Two inbounds backing the service, `appci=jeffy` vs `appci=bob` → key dropped + metric incremented.
6. Inbound tag `kuma.io/protocol=http` → NOT propagated (reserved).
7. Inbound tag `k8s.kuma.io/namespace=foo` → NOT propagated (reserved).
8. Invalid key `BAD!KEY` → skipped + metric incremented; other labels still propagate.
9. Allow-list `["appci"]`; tags include `appci=jeffy` and `team=payments` → only `appci` propagated.

**Cross-DP (`mergeAcrossDataplanes`):**
12. All DPs agree: `team=payments` → propagated.
13. 7 DPs `version=v5`, 3 DPs `version=v6` → `version=v5` (clear majority).
14. 9 DPs `version=v5`, 1 newer DP `version=v6` → `version=v5` (canary safety).
15. 5 DPs `v5`, 5 newer DPs `v6` → `version=v6` (tie, newest wins).
16. 5 DPs no `tier`, 5 DPs `tier=critical` → `tier=critical` (only one value voted).
17. 5 old DPs `team=payments`, 5 new DPs without `team` → `team=payments` (key still defined).
18. After last DP defining `team` is removed → output omits `team` (zero votes for the key).
19. Tied count AND tied creation time → lexicographic value tiebreak.

## Acceptance

- `make check && make test TEST_PKG_LIST=./pkg/core/resources/apis/meshservice/...` passes.
- All 17 test cases above pass.
- Helpers are pure (no side effects beyond logging + metric increment).

## Estimated LOC

~140 lines code + ~240 lines test across 2 files.

## Depends on

Phase 1 (uses `IsReservedLabelKey`).

## Parallel with

Phase 2.

### PR DESCRIPTION
**Pull request by Automaat**
## Motivation

Phase 3 of label propagation: introduce pure, unit-testable helper functions that compute per-DP label contributions and merge them across all dataplanes contributing to a MeshService. Keeping these functions free of I/O side-effects makes them straightforward to test exhaustively and reuse in future phases.

Closes https://github.com/kumahq/kuma/issues/16545

## Implementation information

Two new unexported helpers added to `pkg/core/resources/apis/meshservice/generate/labels.go`:

- `dpContribution` — computes the non-system labels for a single `DataplaneResource` by (1) reaching cross-inbound consensus on inbound tags, (2) overlaying DP resource labels, then (3) validating keys/values and applying the optional allow-list. Drops are counted via the existing `prometheus.Counter` so the metric stays source-agnostic.
- `mergeAcrossDataplanes` — aggregates per-DP contributions into a single map using majority-count semantics; ties broken by newest creation time then lexicographic value order.

The follow-up fix (`fix(meshservice): uniform allow-list drop accounting`) ensures allow-list rejections on inbound tags are counted the same way as DP label rejections, keeping the dropped-label metric consistent regardless of where the label originated.

Both helpers are exported via `export_test.go` for white-box unit tests; 21 table-driven test entries cover edge cases including nil DPs, tied votes, invalid keys/values, and purity (no mutation of inputs).

> Changelog: feat(kuma-cp): support running without inbound tags

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -296,6 +296,8 @@ require (
 	sigs.k8s.io/structured-merge-diff/v6 v6.4.0 // indirect
 )
 
+require github.com/kylelemons/godebug v1.1.0 // indirect
+
```
Comment (bartsmykla) [inline go.mod]: Don't add a standalone `require` block — the existing indirect block at lines 103-297 is the right home for this dep. Run `go mod tidy` and it'll insert this line in alphabetical position automatically.

```suggestion
```

File: pkg/core/resources/apis/meshservice/generate/labels.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+			continue
+		}
+		if errs := validation.IsValidLabelValue(e.value); len(errs) > 0 {
+			metric.Inc()
+			log.V(1).Info("dropping invalid label value", "key", k, "value", e.value, "dp", dp.Meta.GetName())
+			continue
+		}
+		out[k] = e.value
+	}
+
```
Comment (bartsmykla) [inline pkg/core/resources/apis/meshservice/generate/labels.go:78]: Comments read Step 1 → Step 3 → Step 2 in execution order, which forces the reader to jump backwards to make sense of "Step 2" running last. Pick one of:

- renumber the comments to match execution order (consensus build → consensus validate → DP overlay), or
- reorder the code so DP overlay runs first and the numbers line up.

Either is fine, just don't ship the current ordering.

**Review by bartsmykla — COMMENTED** [review]
Two items: the new go.mod `require` block should be merged into the existing indirect block (run `go mod tidy`), and the Step 1/3/2 comment numbering in `labels.go` is confusing.

**Review by bartsmykla — DISMISSED** [review]

**Review by bartsmykla — APPROVED** [review]
